R/stat-density.R

```diff
@@ -112,17 +114,19 @@ StatDensity <- ggproto("StatDensity", Stat,
 compute_density <- function(x, w, from, to, bw = "nrd0", adjust = 1,
                             kernel = "gaussian", n = 512,
                             bounds = c(-Inf, Inf)) {
-  nx <- length(x)
+  nx <- w_sum <- length(x)
   if (is.null(w)) {
     w <- rep(1 / nx, nx)
   } else {
-    w <- w / sum(w)
+    w_sum <- sum(w)
+    w <- w / w_sum
```

Comment:
The function `fit_data_to_bounds()` checks for division by zero but here you're not doing that. I think in weird corner cases with zero weights this could give you different results depending on whether a point is inside the bounds or not.

Reply:
I've been trying to come up with examples that should be valid, but aren't, but I'm not having an easy time of it. If you censor a non-zero weight and are left with all zero weights, the weighted density should reflect that and show a flat line. Currently, it does this, but also passes on a warning from `stats::density()`: 

``` r
devtools::load_all("~/packages/ggplot2/")
#> ℹ Loading ggplot2
df <- data.frame(x = 1:3)

ggplot(df, aes(x)) +
  geom_density(aes(weight = c(1, 0, 0), y = after_stat(wdensity)),
               bounds = c(1.1, Inf))
#> Warning: Some data points are outside of `bounds`. Removing them.
#> Warning in density.default(x, weights = w, bw = bw, adjust = adjust, kernel =
#> kernel, : sum(weights) != 1 -- will not get true density
```

![](https://i.imgur.com/RF5bK10.png)<!-- -->

<sup>Created on 2023-04-11 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

I'm wondering if there is something to be improved here.

Reply:
Ok. As far as I'm concerned you can handle it however you see fit, I just wanted to call out the inconsistency.